### ISSUE TITLE
gelsd! and gelsy! don't do what their documentation says

### ISSUE THREAD

**Issue by tkoolen** [JuliaLang/julia#22242]
Consider the following code:
```julia
A = rand(3, 3)
b = rand(3)
x1 = A \ b
x2 = copy(b)
x3, r = LinAlg.LAPACK.gelsd!(A, x2, 0.)
```
According to [the `gelsd!` documentation](https://docs.julialang.org/en/stable/stdlib/linalg/?highlight=ldlt#Base.LinAlg.LAPACK.gelsd!), `x2` should be overwritten by the solution, so I would expect that `x1`, `x2`, and `x3` are all the same. However, we only have `isapprox(x1, x3)`; `x2` is still equal to `b` after the call to `gelsd!`.

The same is true for `gelsy!`.

Importantly, this implies that the result vector is being allocated inside of `gelsd!` and `gelsy!`, which is what I'm trying to prevent by calling these functions.

**Comment by KristofferC** [comment]
Yep, there are a lot of allocations going on in that function.

```
newB = [B; zeros($elty, max(0, n - size(B, 1)), size(B, 2))]
s = similar(A, $relty, min(m, n))
.... 
subsetrows(B, newB, n)
```

where

```
subsetrows(X::AbstractVector, Y::AbstractArray, k) = Y[1:k]
subsetrows(X::AbstractMatrix, Y::AbstractArray, k) = Y[1:k, :]
```

also copies.

**Comment by KristofferC** [comment]
I think that is a problem with a lot of the lapack wrappers. They are too high level in the sense that there is no way to avoid the allocations. There should perhaps be a two layer wrapping, one that is pretty much a "naked" wrapper around the ccall with the exception of stuff like length of arrays that can be derived from the input data and then another which also create work arrays needed in the lapack-C functions.

**Comment by yuyichao** [comment]
https://github.com/JuliaLang/julia/issues/16263

**Comment by andreasnoack** [comment]
Yes, I agree that the general issue is #16263. However, the motivation there is mainly to have a way to avoid allocation of workspace arrays whereas this is the solution array. I think some version of https://github.com/JuliaLang/julia/issues/22242#issuecomment-306480232 is the right way to do this. The lowest level wrapper should take all arrays as input arguments and just check that the shapes and values are valid.

**Comment by KristofferC** [comment]
I don't see how the solution array could be stored in `b` when the result doesn't even have to be the same size as `b`.

**Comment by tkoolen** [comment]
Good point. I didn't even consider that because my `A` was square.
